waspls/src/Wasp/Backend/ConcreteParser/Internal.hs

```diff
@@ -0,0 +1,701 @@
+      advance
+      let newNode =
+            SyntaxNode
+              { snodeKind = S.Token expectedKind,
+                snodeWidth = tokenWidth nextToken,
+                snodeChildren = []
+              }
+      pushNode newNode
+    nextToken -> throwParseError nextToken (TokenSet.fromKind expectedKind)
+
+parseEOF :: ParserM ()
+parseEOF =
+  (consumeTrivia >> peek) >>= \case
+    Just nextToken -> do
+      nextTokenOffset <- gets pstateNextTokenOffset
+      advance
+      let nextTokenKind = tokenKind nextToken
+      -- TODO: consider reporting all kinds here? not sure
+      _tokenKinds <- collectUntilEOF
+      throwError $
+        ParseError $
+          Unexpected
+            (Span nextTokenOffset (nextTokenOffset + tokenWidth nextToken))
+            nextTokenKind
+            TokenSet.fromEOF
+      where
+        collectUntilEOF :: ParserM [TokenKind]
+        collectUntilEOF =
+          peek >>= \case
+            Just next -> advance >> ((tokenKind next :) <$> collectUntilEOF)
+            Nothing -> pure []
+    Nothing -> do
+      return ()
+
+parseAs :: SyntaxKind -> TokenKind -> ParserM ()
+parseAs label kind =
+  (consumeTrivia >> peek) >>= \case
+    Just nextToken | tokenKind nextToken == kind -> do
+      advance
+      let newNode =
+            SyntaxNode
+              { snodeKind = label,
+                snodeWidth = tokenWidth nextToken,
+                snodeChildren = []
+              }
+      pushNode newNode
+    nextToken -> throwParseError nextToken (TokenSet.fromKind kind)
+
+parseGroup :: SyntaxKind -> GrammarRule -> ParserM ()
+parseGroup label inner = do
+  consumeTrivia
+  previousNodes <- gets pstateNodes
+  modify (\s -> s {pstateNodes = []})
+  maybeInnerParserError <- tryError $ parseRule inner
+  childNodes <- gets pstateNodes
+  modify (\s -> s {pstateNodes = previousNodes})
+  case maybeInnerParserError of
+    Nothing -> pushGroupNode childNodes
+    Just (ParseError parseError) -> handleInnerParseError childNodes parseError
+    Just (Unwind nextTokenKind) -> handleInnerUnwind childNodes nextTokenKind
+  where
+    handleInnerParseError :: [SyntaxNode] -> ParseError -> ParserM ()
+    handleInnerParseError childNodes parseError = do
+      -- ASSUMPTION: The token errored on has not been consumed
+      --
+      -- (1) Check whether the kind we errored on is in the first `pstateFollowingTokens`
+      -- list.
+      --
+      -- (2) If if it is not in the following stack, consume the error token.
+      --
+      -- (3) Report the error and create the group node
+      --
+      -- (4) If the kind is not in first "TokenSet" in the following stack, start
+      -- unwinding.
+
+      -- (1)
+      followingTokenStack <- gets pstateFollowingTokens
+      let expectedKind = case parseError of
+            UnexpectedEOF _ _ -> Nothing
+            Unexpected _ k _ -> Just k
+      let expectedKindIsFollowingKind = any (expectedKind `TokenSet.member`) followingTokenStack
```

Comment:
So here you want to check if the token we expected when the error happened is among the tokens expected by the following rules, right?
So let's name it differently than, maybe smth like `isExpectedTokenKindAmongFollowingExpectedTokenKinds`? Although I wish we had a better name for "following expected token kinds". But if we want to have it we need to introduce it somewhere earlier and stick with it.
Maybe we could go with name `doesAnyFollowingRuleHandleExpectedToken` or smth like that?

Reply:
Actually, what is this `expectedKind` -> it is not an expected token kind it should be unexpected token kind, right? I mean, it is kind of a token that parsing errored on, right? So why are we calling it `expected`?